docker-start.sh

```diff
@@ -1,14 +1,25 @@
-#!/bin/bash
-set -e
-
-PG_HOST=db
-PG_PORT=5432
-
-echo "Wait for postgres server to be ready..."
-while ! nc -q 1 $PG_HOST $PG_PORT </dev/null; do sleep 1; done
+#!/bin/sh
+set -eu
 
 # Create temporary directories
 mkdir -p tmp/pids
 
-bin/rails db:migrate
+# We need to wait for the database to be ready until we can run migrations.
+# To do this, retry up to 15 times to run the migrations. Once that worked,
+# we can assume that the database is up and ready.
+success=0
+retries=15
+for i in $(seq $retries); do
+    echo "Running migrations…"
+    bin/rails db:migrate && success=1 && break || true
+    echo "Database not yet ready, waiting ($i/$retries)…"
+    sleep 1
+done
+if [ "$success" -eq 1 ]; then
+    echo "Database is up, all migrations applied"
+else
+    echo "Timed out waiting for database"
+    exit 1
+fi
```

Comment:
It's not reliable, since the PostgreSQL image listens on the port already when still initializing the cluster. That means that there might be a race condition with the wait script.

Running the migrations until it works is more reliable, it shows that a connection is actually possible.

Reply:
It's not reliable, since the PostgreSQL image listens on the port already when still initializing the cluster. That means that there might be a race condition with the wait script (if it reports the database as up, when it's not actually ready).

Running the migrations until it works is more reliable, it shows that a connection is actually possible.